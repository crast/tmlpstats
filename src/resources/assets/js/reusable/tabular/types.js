import Immutable from 'immutable'

const baseColumn = Immutable.Record({
    key: null,         // Required; key of the table.
    label: null,       // Basically required: heading.
    sortable: true,   // If true, allow sorting on this column.
    component: null,   // if set, a react component.
    selector: null,    // If set, a selector to get this column's value. Defaults to a generated selector.
    sorter: 'string',  // What to use for sorts
    sortSelector: null,// if set, a selector to get the sort value.

    // Internal, used by us
    _columnClick: null,
    _sort: null
})

export class Column extends baseColumn {
    constructor(input) {
        if (!input.selector) {
            input.selector = makeKeySelector(input.key)
        }
        super(input)
    }
}


function makeKeySelector(key) {
    return (obj) => obj[key]
}

export const TableState = Immutable.Record({
    sort_by: Immutable.List(),
})
