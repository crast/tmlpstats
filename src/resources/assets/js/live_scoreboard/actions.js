import objectPick from 'lodash/pick'

import { scoreboardLoad, redux_scoreboard } from './data'
import Api from '../api'

const loadState = scoreboardLoad.actionCreator()

const DESIRED_SB_KEYS = ['games', 'meta', 'week']

export function getCurrentScores(center) {
    return (dispatch, _, { Api }) => {
        dispatch(loadState('loading'))
        Api.LiveScoreboard.getCurrentScores({center}, function (data) {
            dispatch(updateScoreboard(objectPick(data, DESIRED_SB_KEYS)))
            dispatch(loadState('loaded'))
        })
    }
}

export function updateScoreboard(data) {
    return {
        type: 'live_scoreboard/setAll',
        payload: data
    }
}

export function changeGameField(game, field, value) {
    return {
        type: 'live_scoreboard/setGameValue',
        payload: { game, field, value }
    }
}

export function setGameOp(game, op) {
    return changeGameField(game, 'op', op)
}

export function postGameValue(center, game, type, value) {
    return (dispatch, getState) => {
        const request = { center, game, type, value }
        return Api.LiveScoreboard.setScore(request).then((data) => {
            if (data.games) {
                const existing = getState().live_scoreboard.scoreboard
                dispatch(updateScoreboard(redux_scoreboard.mergeGameUpdates(existing, data.games)))
            }
        })
    }
}

export function submitUpdates(center, game, field, gameValue) {
    return (dispatch, getState) => {
        dispatch(setGameOp(game, 'updating'))
        const makeRevert = () => {
            const generation = generationSelector(getState(), game)
            return ifGeneration(getState, game, () => {
                dispatch(setGameOp(game, 'default'))
            })
        }
        const successHandler = (data) => {
            dispatch(setGameOp(game, 'success'))
            setTimeout(makeRevert(), 5000)
            return data
        }
        const failHandler = () => {
            dispatch(setGameOp(game, 'failed'))
            setTimeout(makeRevert(), 8000)
        }
        return dispatch(postGameValue(center, game, field, gameValue)).then(successHandler, failHandler)
    }
}

export function changeGameFieldPotentialUpdate(center, game, field, gameValue) {
    return (dispatch, getState) => {
        dispatch(changeGameField(game, field, gameValue))

        if (!isNaN(parseInt(gameValue))) {
            const cb = ifGeneration(getState, game, () => {
                dispatch(submitUpdates(center, game, field, gameValue))
            })
            setTimeout(cb, 700)
        }
    }

}

function ifGeneration(getState, game, handler) {
    const generation = generationSelector(getState(), game)
    return () => {
        if (generationSelector(getState(), game) == generation) {
            handler(...arguments)
        }
    }
}

function gameSelector(state, game) {
    return state.live_scoreboard.scoreboard.games[game]
}

function generationSelector(state, game) {
    return gameSelector(state, game)._gen
}
