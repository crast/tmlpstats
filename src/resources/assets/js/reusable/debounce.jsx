// Taken from https://github.com/podefr/react-debounce-render
//
// This code licensed under original MIT license:
// https://github.com/podefr/react-debounce-render/blob/master/LICENSE.md

import React, { Component } from 'react'
import _ from 'lodash'

export default function debounceRender(ComponentToDebounce, ...debounceArgs){
    return class DebouncedContainer extends Component {
        constructor(props) {
            super(props);

            this.state = props;
            this.shouldRender = false;
        }

        componentWillReceiveProps(props) {
            this.shouldRender = false;
            this.updateState(props);
        }

        updateState = _.debounce(props => {
            this.shouldRender = true;
            this.setState(props);
        }, ...debounceArgs);

        shouldComponentUpdate() {
            return this.shouldRender;
        }

        render() {
            return <ComponentToDebounce { ...this.state } />;
        }
    }
}
