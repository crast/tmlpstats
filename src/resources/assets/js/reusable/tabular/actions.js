export const CLICK_COLUMN = 'tabular/click-column'
export const CLICK_ITEM = 'tabular/click-item'
export const OVERRIDE_SORT = 'tabular/override-sort'


// clickColumn is used when we click a column, maybe with a modifier key.
export function clickColumn(table, column, modKey) {
    return {type: CLICK_COLUMN, table, column, modKey}
}

export function overrideSort(table, sorts) {
    return {type: OVERRIDE_SORT, table, sorts}
}
