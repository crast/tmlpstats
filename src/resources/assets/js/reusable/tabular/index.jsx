/**
 The tabular library provides a way to make sortable table components with other features.

 Features include:
    * Efficient sorting clicking on column headings including compound sorts
    * Manages its own state without connecting more reducers
    * Tries to minimize expensive re-renders

 To use it, you first must create a component with buildTable:

const MyTable  = buildTable({
    name: 'unique_slug',
    columns: [
        {key: 'name', label: 'Name', default: 'N/A'}
        {key: 'acc', label: 'Accountability', selector: row => row.accountability.display},
        {key: 'updatedAt', label: 'Updated At', sorter: 'moment',
         headingClasses: ['hidden-print'], component: PrettyDate}
    ]
})

Columns specify the columns in the table. Column specifiers are objects with
the following properties:

|    Property    |    Type   |                    Description                    |       Default        |
| -------------: | --------- | ------------------------------------------------- | -------------------- |
|            key | string    | Required. Must be a unique key for this column.   |                      |
|          label | string    | Required. Heading label.                          |                      |
|       selector | func      | A function to select the value for this column.   | x => x[key]          |
|        default | <any>     | Default value if the result of selector is empty  | undefined            |
|      component | Component | A react component to render this row.             | <internal component> |
|       sortable | bool      | Set to false to disable column sorting            | true                 |
|         sorter | string    | Sort method, one of 'string', 'number', 'moment'. | 'string'             |
|   sortSelector | func      | A selector to find the sort candidate value.      | <selector>           |
| headingClasses | array     | Array of additional CSS classes                   | []                   |


*/
import PropTypes from 'prop-types'
import React from 'react'
import Immutable from 'immutable'
import ImmutablePropTypes from 'react-immutable-proptypes'

import { defaultMemoize, createSelector } from 'reselect'
import { connect } from 'react-redux'

import { Glyphicon, cssClasses } from '../ui_basic'
import { compositeKey } from '../sort-helpers'
import { Column, TableState } from './types'
import { clickColumn } from './actions'


class TabularBase extends React.Component {
    static propTypes = {
        current: PropTypes.instanceOf(TableState),
        data: PropTypes.oneOfType([
            PropTypes.shape({forEach: PropTypes.func}),
            PropTypes.array,
        ]),
        tableClasses: PropTypes.string,
        idExtension: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        columnContext: PropTypes.any
    }

    constructor(props) {
        super(props)
        this._boundColumns = createBoundColumnsSelector(this)
        this.sortedRows = createSortedRowsSelector()
    }

    render() {
        const { props } = this
        let idExtension = props.idExtension? '-' + props.idExtension: ''
        const htmlId = `tbl-${this.config.name}${idExtension}`

        return (
            <table className={props.tableClasses} id={htmlId}>
                {this.renderHeading(htmlId)}
                {this.renderRows()}
            </table>
        )
    }

    boundColumns() {
        return this._boundColumns(this.config.columns, this.props.current.sort_by)
    }

    renderHeading(htmlId) {
        let headings = []
        const boundColumns = this.boundColumns()
        boundColumns.forEach((col) => {
            let ascDesc
            let classes = col.headingClasses || []
            if (col.sortable) {
                classes.push('tabular-sortable')
                if (col._sort) {
                    classes.push(col._sort)
                    // sort-by-alphabet{-alt} and sort-by-attributes{-alt}
                    const icon = 'sort-by-' + (col.sorter == 'string'? 'alphabet' : 'attributes') + (col._sort == 'asc'? '' : '-alt')
                    ascDesc = <Glyphicon icon={icon} />
                }
            }
            headings.push(<th key={col.key} onClick={col._columnClick} className={cssClasses(classes)} aria-controls={htmlId}>{col.label}{ascDesc}</th>)
        })
        return (
            <thead><tr>{headings}</tr></thead>
        )
    }

    renderRows() {
        let rows = []
        const {getPrimaryKey, columns} = this.config
        this.sortedRows(this).forEach((row) => {
            rows.push(<TabularRow key={getPrimaryKey(row)} row={row} columns={columns} columnContext={this.props.columnContext} />)
        })
        return (
            <tbody>{rows}</tbody>
        )
    }

    onColumnClick(colKey, event) {
        event.persist()
        const column = this.config.columns.get(colKey)
        if (column.sortable) {
            this.props.dispatch(clickColumn(this.config.name, colKey, event.ctrlKey || event.shiftKey || event.altKey))
        } else {
            console.log('clicked on un-sortable column ' + colKey)
        }
    }
}

function createSortedRowsSelector() {
    const sorterForColumns = defaultMemoize((columns, sort_by) => {
        const sortKeys = sort_by.map((colSort) => {
            const col = columns.get(colSort.column)
            return [(col.sortSelector || col.selector), col.sorter, colSort.direction]
        })
        return compositeKey(sortKeys.toJS())
    })
    return createSelector(
        (c) => c.props.data,
        (c) => c.props.current.sort_by,
        (c) => c.config.columns,
        (rows, sort_by, columns) => {
            if (!sort_by || !sort_by.size) {
                return rows
            }
            const sortFunc = sorterForColumns(columns, sort_by)
            return rows.slice().sort(sortFunc)
        }
    )
}

function createBoundColumnsSelector(component) {
    return defaultMemoize((columns, sort_by) => {
        const keyed_sort_by = Immutable.Map(sort_by.map(x => [x.column, x.direction]))
        return columns.map((col) => {
            return col
                .set('_columnClick', component.onColumnClick.bind(component, col.key))
                .set('_sort', keyed_sort_by.get(col.key))
        })
    })
}

class TabularRow extends React.PureComponent {
    static propTypes = {
        row: PropTypes.any.isRequired,
        columns: ImmutablePropTypes.orderedMapOf(
            PropTypes.instanceOf(Column),  // Value type
            PropTypes.string // Key type
        ),
        columnContext: PropTypes.any
    }

    render() {
        const { row, columns } = this.props
        let cols = []
        columns.forEach((col) => {
            const Component = col.component
            const data = col.selector(row) || col.default
            const rendered = Component? (<Component key={col.key} data={data} columnContext={this.props.columnContext} />) : <td key={col.key}>{data}</td>
            cols.push(rendered)
        })
        return <tr>{cols}</tr>
    }
}

/**
 * Create a table component.
 * @param  object options An object with the following properties:
 *
 *            name | string | Required. An identifier for this table for property storage.
 *         columns | array  | Required. An array of column specifier objects. See top of module for docs.
 *   getPrimaryKey | func   | A selector to get the 'primary key' of a row. defaults to x => x.id
 *     displayName | string | DisplayName primarily for React debugging use.
 *
 * @return React.Component A component used to render a table.
 */
export function buildTable(options) {
    const tableName = options.name
    const defaultState = options.defaultState || defaultTableState
    options.getPrimaryKey = options.getPrimaryKey || defaultPrimaryKeyFunc
    options.columns = Immutable.Seq.Indexed(options.columns)
        .map((c) => {
            return (c.set) ? c : new Column(c)
        })
        .toKeyedSeq()
        .mapEntries((x) => [x[1].key, x[1]])
        .toOrderedMap()

    class MyTable extends TabularBase {
        config = options
    }

    MyTable.defaultProps = {
        tableClasses: options.tableClasses || 'table'
    }

    MyTable.displayName = options.displayName || tableName

    if (options.noConnect) {
        return MyTable
    }

    function mapState(state) {
        return {
            current: state.tabular.tables.get(tableName, defaultState)
        }
    }

    return connect(mapState)(MyTable)
}

const defaultTableState = new TableState({})
const defaultPrimaryKeyFunc = x => x.id
