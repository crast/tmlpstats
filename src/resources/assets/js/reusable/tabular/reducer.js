// This is the reducer for Tabular.
import Immutable from 'immutable'

import { CLICK_COLUMN, OVERRIDE_SORT } from './actions'
import { TableState, ColumnSort, adaptColumnSorts } from './types'


const R = new Immutable.Record({
    tables: Immutable.Map()
})
const startState = new R()

const EMPTY_TABLE_STATE = new TableState({})
const LAST = -1 // The -1'th element is syntactic sugar for last element in an immutable list.


function reducer(state=startState, action) {
    if (action.type.startsWith('tabular/')) {
        let current = state.tables.get(action.table, EMPTY_TABLE_STATE)
        if (action.type === CLICK_COLUMN) {
            let sortBy = current.sort_by
            if (sortBy.size && sortBy.get(LAST).column == action.column) {
                // If the last item in the sort is the column just clicked, simply flip asc/desc.
                const v = sortBy.get(LAST)
                sortBy = sortBy.set(LAST, v.set('direction', v.direction == 'asc'? 'desc' : 'asc'))
            } else {
                const newSort = new ColumnSort({column: action.column})
                if (action.modKey) {
                    // If a modifier key is pressed, then add to a compound sort.
                    sortBy = sortBy.push(newSort)
                } else {
                    // For a click without modifier keys, replace the whole list with a clean one.
                    sortBy = Immutable.List.of(newSort)
                }
            }
            current = current.set('sort_by', sortBy)
        } else if (action.type === OVERRIDE_SORT) {
            // override sort is a special action not given by our own component, instead done when
            // there is an action maybe to do a custom set of sorts.
            current = current.set('sort_by', adaptColumnSorts(action.sorts))
        }
        return state.setIn(['tables', action.table], current)
    }
    return state
}

export default reducer
