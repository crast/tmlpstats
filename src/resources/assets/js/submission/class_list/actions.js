import { actions as formActions } from 'react-redux-form'

import { bestErrorValue } from '../../reusable/ajax_utils'
import { classListLoad, teamMembersCollection, weeklyReportingData, weeklyReportingSave } from './data'
import { TEAM_MEMBERS_COLLECTION_FORM_KEY, TEAM_MEMBER_FORM_KEY, WEEKLY_REPORTING_CHANGE } from './reducers'
import Api from '../../api'

export const loadState = classListLoad.actionCreator()
const weeklySaveState = weeklyReportingSave.actionCreator()

export function loadClassList(centerId, reportingDate) {
    return (dispatch, _, { Api }) => {
        dispatch(loadState('loading'))
        return Api.TeamMember.allForCenter({
            center: centerId,
            reportingDate: reportingDate,
            includeInProgress: true
        }).done((data) => {
            dispatch(initializeClassList(data))
        }).fail((jqXHR, textStatus) => {
            dispatch(loadState({error: bestErrorValue(jqXHR, textStatus)}))
        })
    }
}

function initializeClassList(data) {
    return (dispatch) => {
        // Re-format the collection as a key-ordered collection
        data = teamMembersCollection.ensureCollection(data)
        dispatch(formActions.load(TEAM_MEMBERS_COLLECTION_FORM_KEY, data))
        dispatch(loadState('loaded'))
    }
}

export function chooseTeamMember(data) {
    console.log('load', TEAM_MEMBER_FORM_KEY, data)
    return formActions.load(TEAM_MEMBER_FORM_KEY, data)
}

export function weeklyReportingUpdated(teamMemberId) {
    return weeklyReportingData.mark(teamMemberId)
}

export function weeklyReportingSubmit(center, reportingDate, tracking, rawData) {
    var updates = []
    for (var id in tracking.changed) {
        let {gitw, tdo} = rawData[id]
        updates.push({id, gitw, tdo})
    }
    console.log('updates', updates)

    return (dispatch) => {
        dispatch(weeklyReportingData.beginWork())
        dispatch(weeklySaveState('loading'))

        const success = (data) => {
            dispatch(weeklySaveState('loaded'))
            setTimeout(() => {
                dispatch(weeklyReportingData.endWork())
                dispatch(weeklySaveState('new'))
            }, 3000)
            return data
        }
        const fail = (err) => {
            dispatch(weeklySaveState({error: err.error || err}))
        }

        return Api.TeamMember.bulkStashWeeklyReporting({center, reportingDate, updates}).then(success, fail)
    }
}
