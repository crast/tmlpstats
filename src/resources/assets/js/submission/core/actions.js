import { coreInit, centerQuarterData, cqResponse } from './data'
import { objectAssign } from '../../reusable/ponyfill'
import { bestErrorValue } from '../../reusable/ajax_utils'
import Api from '../../api'
import { normalize } from 'normalizr'

export const initState = coreInit.actionCreator()

export function initSubmission(centerId, reportingDate) {
    return (dispatch) => {
        dispatch(initState('loading'))
        return Api.SubmissionCore.initSubmission({
            center: centerId,
            reportingDate: reportingDate
        }).then((data) => {
            dispatch(setSubmissionLookups(data))
        }).catch((jqXHR, textStatus) => {
            dispatch(initState({error: bestErrorValue(jqXHR, textStatus)}))
        })
    }
}

export function setSubmissionLookups(data) {
    return (dispatch) => {
        const lookups = objectAssign({}, data.lookups)

        // yuck, but works for now while we're remapping
        const n = normalize(data, cqResponse)
        lookups.validRegQuarters = n.entities.c[n.result].validRegQuarters
        dispatch(centerQuarterData.replaceItems(n.entities.quarters))

        dispatch({
            type: 'core/setSubmissionLookups',
            payload: lookups
        })
        dispatch(initState('loaded'))
    }
}

export function setReportingDate(reportingDate) {
    return {
        type: 'submission.setReportingDate',
        payload: reportingDate
    }
}
