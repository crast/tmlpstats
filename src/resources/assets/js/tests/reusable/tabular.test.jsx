import React from 'react'
import renderer from 'react-test-renderer'
import { Provider, newTestingStore } from '../testing-store'

import * as actions from '../../reusable/tabular/actions'
import reducer from '../../reusable/tabular/reducer'
import { ColumnSort } from '../../reusable/tabular/types'
import { buildTable } from '../../reusable/tabular'

describe('Reducer', () => {
    const TABLE_NAME = 'mytable_x'

    function getTable(result) {
        return result.tables.get(TABLE_NAME)
    }

    test('Defaults', () => {
        expect(reducer(undefined, {type: 'ignoreme'})).toBeDefined()
        expect(reducer(undefined, actions.clickColumn(TABLE_NAME, 'abc'))).toBeDefined()
        expect(reducer(undefined, {type: 'tabular/bogus'})).toBeDefined()
    })

    describe('Column Click', () => {
        it('Should assign a sort column when there is none', () => {
            const result = reducer(undefined, actions.clickColumn(TABLE_NAME, 'abc'))
            expect(result.tables.size).toBe(1)
            const t = getTable(result)
            expect(t.sort_by.size).toBe(1)
            const firstSort = t.sort_by.get(0)
            expect(firstSort.column).toBe('abc')
            expect(firstSort.direction).toBe('asc')
        })

        const singleSort = reducer(undefined, actions.clickColumn(TABLE_NAME, 'def'))

        it('Should flip sort on double-sort of same column', () => {
            const r1 = reducer(singleSort, actions.clickColumn(TABLE_NAME, 'def'))
            const t1 = getTable(r1)
            expect(t1.sort_by.size).toBe(1)
            expect(t1.sort_by.get(0).direction).toBe('desc')
            const r2 = reducer(r1, actions.clickColumn(TABLE_NAME, 'def', true))
            const t2 = getTable(r2)
            expect(t2.sort_by.size).toBe(1)
            expect(t2.sort_by.get(0).direction).toBe('asc')
        })

        it('Should replace on different columnn click', () => {
            const r1 = reducer(singleSort, actions.clickColumn(TABLE_NAME, 'ghi'))
            const t = getTable(r1)
            expect(t.sort_by.size).toBe(1)
            expect(t.sort_by.get(0).column).toBe('ghi')
            expect(t.sort_by.get(0).direction).toBe('asc')
        })

        it('Should compound sort on different columnn click', () => {
            const r1 = reducer(singleSort, actions.clickColumn(TABLE_NAME, 'ghi', true))
            const t = getTable(r1)
            expect(t.sort_by.size).toBe(2)
            expect(t.sort_by.get(0).column).toBe('def')
            expect(t.sort_by.get(1).column).toBe('ghi')
        })
    })

    describe('Replace Sorts', () => {
        const initial = reducer(undefined, actions.clickColumn(TABLE_NAME, 'ignoreme'))

        it('Should replace sorts', () => {
            const input = [
                {'column': 'aaa'},
                {'column': 'bbb', direction: 'desc'},
                new ColumnSort({column: 'ccc', direction: 'desc'})
            ]
            const r = reducer(initial, actions.overrideSort(TABLE_NAME, input))
            const { sort_by } = getTable(r)
            expect(sort_by.size).toBe(3)
            expect(sort_by.map(x => [x.column, x.direction]).toJS()).toEqual([
                ['aaa', 'asc'], ['bbb', 'desc'], ['ccc', 'desc']
            ])
            expect(sort_by.get(1)).toBeInstanceOf(ColumnSort)
            expect(sort_by.get(2)).toBe(input[2]) // Record should be unmodified

        })

    })
})

describe('Tables', () => {
    const store = newTestingStore()

    const NAME = 'abc'
    const MyTable = buildTable({
        name: NAME,
        columns: [
            {key: 'name', label: 'Name'},
            {key: 'uName', label: 'Upper Name', selector: x => x.name.toUpperCase()},
            {key: 'age', sorter: 'number'}
        ]
    })

    const data1 = [
        {id: 1, name: 'Bob', age: 55},
        {id: 2, name: 'Mary', age: 49},
        {id: 3, name: 'Joan', age: 38}
    ]

    it('Should Render Snapshot', () => {
        const tree = renderer.create(
            <Provider store={store}>
                <MyTable data={data1} />
            </Provider>
        ).toJSON()

        expect(tree).toMatchSnapshot()
    })


    const getHeadings = tree => tree.children[0].children[0].children // thead -> tr -> children
    const getTbodyRows = tree => tree.children[1].children // [thead, tbody] .. children
    const getNameOnly = row => row.children[0].children[0] // first td -> text node

    it('Should respond to sorts', () => {
        function renderAndFetchNames() {
            const tree = renderer.create(
                <Provider store={store}>
                    <MyTable data={data1} />
                </Provider>
            ).toJSON()

            const rows = getTbodyRows(tree)
            const names = rows.map(getNameOnly)
            return {tree, rows, names}
        }

        store.dispatch(actions.clickColumn(NAME, 'name'))
        expect(renderAndFetchNames().names).toEqual(['Bob', 'Joan', 'Mary'])

        store.dispatch(actions.clickColumn(NAME, 'name')) // Flips to descending
        const { tree: tree1, names: names1 } = renderAndFetchNames()
        expect(names1).toEqual(['Mary', 'Joan', 'Bob'])

        // Instead of using store.dispatch let's use an action handler simulating a real click
        getHeadings(tree1)[2].props.onClick({persist: function() {}})
        expect(renderAndFetchNames().names).toEqual(['Joan', 'Mary', 'Bob'])

    })

})
