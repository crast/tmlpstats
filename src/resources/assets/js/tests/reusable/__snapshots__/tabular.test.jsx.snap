// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tables Should Render Snapshot 1`] = `
<table
  className="table"
  id="tbl-abc"
>
  <thead>
    <tr>
      <th
        aria-controls="tbl-abc"
        className="tabular-sortable"
        onClick={[Function]}
      >
        Name
      </th>
      <th
        aria-controls="tbl-abc"
        className="tabular-sortable"
        onClick={[Function]}
      >
        Upper Name
      </th>
      <th
        aria-controls="tbl-abc"
        className="tabular-sortable"
        onClick={[Function]}
      />
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Bob
      </td>
      <td>
        BOB
      </td>
      <td>
        55
      </td>
    </tr>
    <tr>
      <td>
        Mary
      </td>
      <td>
        MARY
      </td>
      <td>
        49
      </td>
    </tr>
    <tr>
      <td>
        Joan
      </td>
      <td>
        JOAN
      </td>
      <td>
        38
      </td>
    </tr>
  </tbody>
</table>
`;
